common/src/comptr.rs

```diff
@@ -2,55 +2,51 @@
 // actually possessing the entire Human struct, just
 // an interface pointer.
 use super::*;
-use std::os::raw::c_void;
 
 use std::ops::Deref;
 use std::ops::DerefMut;
+use std::ptr::NonNull;
 
-pub struct ComPtr<'a, T:'a> {
-    // Interface pointer for interface T as a box.
-    raw_ptr: &'a mut T
+pub struct ComPtr<T> {
+    ptr: NonNull<T>,
 }
 
-impl<T> Deref for ComPtr<'_, T> {
+impl<T> Deref for ComPtr<T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
-        self.raw_ptr
+        unsafe { self.ptr.as_ref() }
     }
 }
 
-impl<T> DerefMut for ComPtr<'_, T> {
+impl<T> DerefMut for ComPtr<T> {
     fn deref_mut(&mut self) -> &mut Self::Target {
-        self.raw_ptr
+        unsafe { self.ptr.as_mut() }
     }
 }
 
-impl<T> ComPtr<'_, T> {
-    pub fn create_instance(rclsid: REFCLSID, pUnkOuter: *mut IUnknown, dwClsContext: DWORD, riid: REFIID) -> Self {
-        let mut p = std::ptr::null_mut::<c_void>();
-
-        unsafe {
-            // let hr = CoCreateInstance(rclsid, std::ptr::null_mut() as *mut IUnknown, dwClsContext, riid, &mut p);
-            let hr = CoGetClassObject(rclsid, CLSCTX_INPROC_SERVER, std::ptr::null_mut::<c_void>(), riid, &mut p as *mut LPVOID);
-            if failed(hr) {
-                panic!("Failed to create instance!");
-            }
-        
-            let raw_ptr = &mut *(p as *mut T);
-
-            ComPtr {
-                raw_ptr
-            }
-        }
+impl<T> ComPtr<T> {
+    /// NonNull<T> must be safely convertable to *mut RawIUnknown
+    pub unsafe fn new(ptr: NonNull<T>) -> Self {

```

Comment:
change this to:
```rust 
pub unsafe fn new_unchecked(ptr: NonNull<T>) -> Self {
```
And then introduce:
```rust
pub fn new(ptr: Box<IUnknown>) -> Self {
```

Or better yet, just use an unsafe trait to make the gurantee...